File: matlab/src/+pme/filters.m
function [DBf, info, mask] = filters(DB, cfg, layout)
%PME.FILTERS Apply column-wise filters in a defined chain.
%
% Order:
%   1) remove_nan
%   2) goal-oriented (hard physical constraints)
%   3) IQR row-by-row outlier filter (legacy behaviour)
%
% Returns:
%   DBf  = DB(:,mask)
%   mask = logical(1,Nsamp) on the ORIGINAL DB columns
%   info = struct with step-by-step + cumulative counts
%
% Logging:
%   [filters][nan]  kept step / in, cumulative kept / total
%   [filters][goal] ...
%   [filters][iqr]  ...
%   [filters] total kept / total

Ns = size(DB,2);
mask = true(1,Ns);
info = struct();
info.Ns_total = Ns;

%% ---------------- Defaults ----------------
if ~isfield(cfg,'filters') || isempty(cfg.filters)
    cfg.filters = struct();
end

if ~isfield(cfg.filters,'remove_nan'), cfg.filters.remove_nan = true; end

if ~isfield(cfg.filters,'goal') || isempty(cfg.filters.goal)
    cfg.filters.goal = struct('enable',false);
end

if ~isfield(cfg.filters,'iqr') || isempty(cfg.filters.iqr)
    cfg.filters.iqr = struct('enable',false);
end
if ~isfield(cfg.filters.iqr,'k'), cfg.filters.iqr.k = 3; end
if ~isfield(cfg.filters.iqr,'start_row'), cfg.filters.iqr.start_row = "after_geom_vars"; end

% Debug switches (only prints; does not affect logic)
if ~isfield(cfg.filters,'debug'), cfg.filters.debug = struct(); end
if ~isfield(cfg.filters.debug,'goal_rows'), cfg.filters.debug.goal_rows = false; end
if ~isfield(cfg.filters.debug,'iqr_rows'),  cfg.filters.debug.iqr_rows  = false; end

% Helper for printing
print_step = @(tag, kept_step, in_step, kept_cum) ...
    fprintf('[filters][%s] kept %d / %d (cumulative %d / %d)\n', tag, kept_step, in_step, kept_cum, Ns);

% Helper: force keep baseline_col (interpreted as "baseline sample column")
function force_keep_baseline()
    if isfield(cfg,'baseline_col') && ~isempty(cfg.baseline_col)
        bc = double(cfg.baseline_col);
        if bc >= 1 && bc <= Ns
            mask(bc) = true;
        end
    end
end

%% =====================================================
%% (1) NaN filter
%% =====================================================
if cfg.filters.remove_nan
    in_step = sum(mask);
    ok = all(~isnan(DB(:,mask)), 1);   % evaluate only currently kept cols (faster & clearer)
    kept_step = sum(ok);

    % Update global mask
    idx = find(mask);
    mask(idx(~ok)) = false;

    % Never drop baseline_col
    force_keep_baseline();

    info.nan_enable   = true;
    info.nan_in       = in_step;
    info.nan_kept     = kept_step;
    info.nan_dropped  = in_step - kept_step;

    print_step('nan', kept_step, in_step, sum(mask));
else
    info.nan_enable   = false;
    info.nan_in       = sum(mask);
    info.nan_kept     = sum(mask);
    info.nan_dropped  = 0;

    % Still enforce baseline
    force_keep_baseline();

    fprintf('[filters][ nan] disabled (cumulative %d / %d)\n', sum(mask), Ns);
end

%% =====================================================
%% (2) Goal-oriented filter (physical constraints)
%% =====================================================
g = cfg.filters.goal;

if isfield(g,'enable') && g.enable

    in_step = sum(mask);

    if ~isfield(layout,'C') || layout.C.nRows == 0
        warning('[filters][goal] enabled but scalar block C is empty. Skipping goal.');
        info.goal_enable  = true;
        info.goal_skipped = true;
        info.goal_in      = in_step;
        info.goal_kept    = in_step;
        info.goal_dropped = 0;

        % Still enforce baseline
        force_keep_baseline();

        fprintf('[filters][goal] skipped (C empty), cumulative %d / %d\n', sum(mask), Ns);

    else
        info.goal_enable  = true;
        info.goal_skipped = false;

        if ~isfield(g,'baseline_col') || isempty(g.baseline_col)
            g.baseline_col = cfg.baseline_col;
        end
        baseline_col = double(g.baseline_col);

        if ~isfield(g,'metrics') || isempty(g.metrics)
            warning('[filters][goal] enabled but cfg.filters.goal.metrics is empty. Skipping goal.');
            info.goal_skipped = true;
            info.goal_in      = in_step;
            info.goal_kept    = in_step;
            info.goal_dropped = 0;

            % Still enforce baseline
            force_keep_baseline();

            fprintf('[filters][goal] skipped (metrics empty), cumulative %d / %d\n', sum(mask), Ns);

        else
            % Work on current DB slice
            idxKept = find(mask);
            DBcur = DB(:, idxKept);

            keep = true(1, size(DBcur,2));

            metrics = g.metrics;

            for i = 1:numel(metrics)

                if iscell(metrics)
                    m = metrics{i};
                else
                    m = metrics(i);
                end

                if isfield(m,'row') && ~isempty(m.row)
                    r = double(m.row);
                elseif isfield(m,'c_offset') && ~isempty(m.c_offset)
                    r = double(layout.C.rows(1) + m.c_offset - 1);
                else
                    disp(m);
                    error('[filters][goal] metric #%d must define row or c_offset.', i);
                end

                % Basic safety
                if r < 1 || r > size(DBcur,1)
                    error('[filters][goal] metric #%d resolved row r=%d out of bounds (1..%d).', i, r, size(DBcur,1));
                end
                if baseline_col < 1 || baseline_col > size(DBcur,2)
                    error('[filters][goal] baseline_col=%d out of bounds (1..%d).', baseline_col, size(DBcur,2));
                end

                base = DBcur(r, baseline_col);
                vals = DBcur(r, :);

                % --- DEBUG print: show which row is checked + 2nd column value ---
                if isfield(cfg.filters,'debug') && isfield(cfg.filters.debug,'goal_rows') && cfg.filters.debug.goal_rows
                    v2_cur = NaN;
                    if size(DBcur,2) >= 2, v2_cur = DBcur(r,2); end

                    v2_orig = NaN;
                    if size(DB,2) >= 2, v2_orig = DB(r,2); end

                    fprintf('[filters][goal][debug] metric %d: rule=%s, r=%d, baseline_col=%d, base=%.6g, DBcur(r,2)=%.6g, DB(r,2)=%.6g\n', ...
                        i, string(m.rule), r, baseline_col, base, v2_cur, v2_orig);
                end

                switch lower(string(m.rule))
                    case "leq_baseline"
                        fac = 1;
                        if isfield(m,'factor') && ~isempty(m.factor), fac = m.factor; end
                        keep = keep & (vals <= fac*base);

                    case "geq_baseline"
                        fac = 1;
                        if isfield(m,'factor') && ~isempty(m.factor), fac = m.factor; end
                        keep = keep & (vals >= fac*base);

                    case "between_baseline"
                        if ~isfield(m,'min') || isempty(m.min) || ~isfield(m,'max') || isempty(m.max)
                            error('[filters][goal] between_baseline requires keys "min" and "max".');
                        end
                        keep = keep & (vals >= m.min*base) & (vals <= m.max*base);

                    case "positive"
                        keep = keep & (vals > 0);

                    case "nonnegative"
                        keep = keep & (vals >= 0);

                    otherwise
                        error('[filters][goal] unknown goal rule: %s', string(m.rule));
                end
            end

            kept_step = sum(keep);

            % Update global mask: drop those not kept
            drop_local = ~keep;
            mask(idxKept(drop_local)) = false;

            % Never drop baseline_col
            force_keep_baseline();

            info.goal_in      = in_step;
            info.goal_kept    = kept_step;
            info.goal_dropped = in_step - kept_step;

            print_step('goal', kept_step, in_step, sum(mask));
        end
    end

else
    info.goal_enable  = false;
    info.goal_skipped = false;
    info.goal_in      = sum(mask);
    info.goal_kept    = sum(mask);
    info.goal_dropped = 0;

    % Still enforce baseline
    force_keep_baseline();

    fprintf('[filters][goal] disabled (cumulative %d / %d)\n', sum(mask), Ns);
end

%% =====================================================
%% (3) IQR row-by-row outlier filter (legacy)
%% =====================================================
iq = cfg.filters.iqr;

if isfield(iq,'enable') && iq.enable

    in_step = sum(mask);

    % Decide start row r0
    if ischar(iq.start_row) || isstring(iq.start_row)

        switch lower(string(iq.start_row))

            case "after_geom_vars"
                % legacy: idx = K*Jdir + Mbase
                K = sum([cfg.geom.patches.K]);  % multipatch supported (sum)
                idx = K * cfg.geom.Jdir + cfg.vars.Mbase;
                r0 = idx + 1;

            case "after_layout_ubase"
                % after D + Ubase (Mbase)
                r0 = layout.D.nRows + cfg.vars.Mbase + 1;

            case "after_c_block_start"
                % start IQR only on scalars C
                if ~isfield(layout,'C') || layout.C.nRows==0
                    warning('[filters][ iqr] start_row=after_c_block_start but C empty. Skipping IQR.');
                    r0 = [];
                else
                    r0 = layout.C.rows(1);
                end

            otherwise
                error('[filters][ iqr] unknown start_row keyword: %s', string(iq.start_row));
        end

    else
        r0 = double(iq.start_row);
    end

    if isempty(r0)
        info.iqr_enable   = true;
        info.iqr_skipped  = true;
        info.iqr_in       = in_step;
        info.iqr_kept     = in_step;
        info.iqr_dropped  = 0;

        % Still enforce baseline
        force_keep_baseline();

        fprintf('[filters][ iqr] skipped (no valid start row), cumulative %d / %d\n', sum(mask), Ns);

    else
        info.iqr_enable   = true;
        info.iqr_skipped  = false;

        % Work on current DB slice
        idxKept = find(mask);
        DBcur = DB(:, idxKept);

        r0 = max(1, min(r0, size(DBcur,1)));
        k = iq.k;

        % --- DEBUG print: show IQR row range actually used ---
        if isfield(cfg.filters,'debug') && isfield(cfg.filters.debug,'iqr_rows') && cfg.filters.debug.iqr_rows
            fprintf('[filters][ iqr][debug] applying IQR from r0=%d to rEnd=%d (k=%g)\n', r0, size(DBcur,1), k);
        end

        valid_cols = true(1, size(DBcur,2));

        for r = r0:size(DBcur,1)
            row_data = DBcur(r,:);

            Q1 = prctile(row_data,25);
            Q3 = prctile(row_data,75);
            IQR = Q3 - Q1;

            lowerBound = Q1 - k*IQR;
            upperBound = Q3 + k*IQR;

            valid_cols = valid_cols & (row_data >= lowerBound) & (row_data <= upperBound);
        end

        kept_step = sum(valid_cols);

        % Update global mask
        mask(idxKept(~valid_cols)) = false;

        % Never drop baseline_col
        force_keep_baseline();

        info.iqr_in            = in_step;
        info.iqr_kept          = kept_step;
        info.iqr_dropped       = in_step - kept_step;
        info.iqr_k             = k;
        info.iqr_start_row_abs = r0;

        print_step('iqr', kept_step, in_step, sum(mask));
    end

else
    info.iqr_enable   = false;
    info.iqr_skipped  = false;
    info.iqr_in       = sum(mask);
    info.iqr_kept     = sum(mask);
    info.iqr_dropped  = 0;

    % Still enforce baseline
    force_keep_baseline();

    fprintf('[filters][ iqr] disabled (cumulative %d / %d)\n', sum(mask), Ns);
end

%% ---------------- Final baseline check ----------------
if isfield(cfg,'baseline_col') && ~isempty(cfg.baseline_col)
    bc = double(cfg.baseline_col);
    if bc >= 1 && bc <= Ns
        if ~mask(bc)
            error('[filters] baseline_col=%d was dropped (this must never happen).', bc);
        end
    end
end

%% =====================================================
% Final
DBf = DB(:, mask);

info.total_kept    = sum(mask);
info.total_dropped = Ns - info.total_kept;

fprintf('[filters] total kept %d / %d samples\n', info.total_kept, Ns);
fprintf('');

end

File: python/src/pme_toolkit/filters.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import numpy as np
from numpy.typing import NDArray


Array = NDArray[np.float64]
BoolArray = NDArray[np.bool_]


@dataclass
class FilterResult:
    db_used: Array
    info: dict[str, Any]
    mask: BoolArray


def _cfg_filters_defaults(cfg: dict[str, Any]) -> dict[str, Any]:
    out = dict(cfg)
    filters = dict(out.get("filters", {}) or {})

    if "remove_nan" not in filters:
        filters["remove_nan"] = True

    goal = dict(filters.get("goal", {}) or {})
    if "enable" not in goal:
        goal["enable"] = False
    filters["goal"] = goal

    iqr = dict(filters.get("iqr", {}) or {})
    if "enable" not in iqr:
        iqr["enable"] = False
    if "k" not in iqr:
        iqr["k"] = 3.0
    if "start_row" not in iqr:
        iqr["start_row"] = "after_geom_vars"
    filters["iqr"] = iqr

    debug = dict(filters.get("debug", {}) or {})
    if "goal_rows" not in debug:
        debug["goal_rows"] = False
    if "iqr_rows" not in debug:
        debug["iqr_rows"] = False
    filters["debug"] = debug

    out["filters"] = filters
    return out


def _baseline_col_zero_based(cfg: dict[str, Any], ns: int) -> int:
    """
    MATLAB uses 1-based baseline_col.
    Python internally uses 0-based indexing.
    """
    bc = int(cfg.get("baseline_col", 1))
    if 1 <= bc <= ns:
        return bc - 1
    if 0 <= bc < ns:
        return bc
    raise ValueError(f"baseline_col={bc} out of bounds for {ns} samples")


def _force_keep_baseline(mask: BoolArray, cfg: dict[str, Any]) -> None:
    ns = mask.size
    bc = _baseline_col_zero_based(cfg, ns)
    mask[bc] = True


def _print_step(tag: str, kept_step: int, in_step: int, kept_cum: int, ns_total: int) -> None:
    print(f"[filters][{tag}] kept {kept_step} / {in_step} (cumulative {kept_cum} / {ns_total})")


def _resolve_goal_row(metric: dict[str, Any], layout: dict[str, Any]) -> int:
    """
    Return 0-based absolute row index in DB.

    JSON convention:
    - metric.row is 0-based absolute DB row
    - metric.c_offset is 0-based offset inside scalar block C

    MATLAB path:
    read_case_json.m converts JSON 0-based -> MATLAB 1-based first,
    then filters.m uses:
        r = layout.C.rows(1) + m.c_offset - 1

    In Python we stay 0-based throughout, so:
        r = layout.C.rows.start + c_offset
    """
    if "row" in metric and metric["row"] not in (None, ""):
        return int(metric["row"])

    if "c_offset" in metric and metric["c_offset"] not in (None, ""):
        c_rows = layout.get("C", {}).get("rows")
        if c_rows is None:
            raise ValueError("[filters][goal] c_offset specified but layout.C.rows is missing")
        return int(c_rows.start) + int(metric["c_offset"])

    raise ValueError("[filters][goal] each metric must define 'row' or 'c_offset'")

def _goal_keep_mask(
    db_current: Array,
    cfg: dict[str, Any],
    layout: dict[str, Any],
) -> BoolArray:
    """
    Goal-oriented hard constraints on currently kept columns.
    Port of the relevant MATLAB logic from filters.m.
    """
    filters_cfg = cfg["filters"]
    goal_cfg = dict(filters_cfg.get("goal", {}) or {})
    debug_cfg = dict(filters_cfg.get("debug", {}) or {})

    metrics = goal_cfg.get("metrics", [])
    if not metrics:
        raise ValueError("[filters][goal] enabled but cfg.filters.goal.metrics is empty")

    keep = np.ones(db_current.shape[1], dtype=bool)

    baseline_col = goal_cfg.get("baseline_col", cfg.get("baseline_col", 1))
    baseline_col = int(baseline_col)
    if 1 <= baseline_col <= db_current.shape[1]:
        bc = baseline_col - 1
    elif 0 <= baseline_col < db_current.shape[1]:
        bc = baseline_col
    else:
        raise ValueError(
            f"[filters][goal] baseline_col={baseline_col} out of bounds "
            f"(1..{db_current.shape[1]})"
        )

    if isinstance(metrics, dict):
        metrics = [metrics]

    for i, metric in enumerate(metrics, start=1):
        m = dict(metric)
        r = _resolve_goal_row(m, layout)

        if r < 0 or r >= db_current.shape[0]:
            raise ValueError(
                f"[filters][goal] metric #{i} resolved row r={r+1} out of bounds "
                f"(1..{db_current.shape[0]})"
            )

        base = float(db_current[r, bc])
        vals = db_current[r, :]

        if bool(debug_cfg.get("goal_rows", False)):
            v2_cur = float(db_current[r, 1]) if db_current.shape[1] >= 2 else np.nan
            print(
                f"[filters][goal][debug] metric {i}: "
                f"rule={str(m.get('rule',''))}, r={r+1}, baseline_col={bc+1}, "
                f"base={base:.6g}, DBcur(r,2)={v2_cur:.6g}"
            )

        rule = str(m.get("rule", "")).lower()

        if rule == "leq_baseline":
            fac = float(m.get("factor", 1.0))
            keep &= vals <= fac * base

        elif rule == "geq_baseline":
            fac = float(m.get("factor", 1.0))
            keep &= vals >= fac * base

        elif rule == "between_baseline":
            if "min" not in m or "max" not in m:
                raise ValueError("[filters][goal] between_baseline requires 'min' and 'max'")
            keep &= (vals >= float(m["min"]) * base) & (vals <= float(m["max"]) * base)

        elif rule == "positive":
            keep &= vals > 0.0

        elif rule == "nonnegative":
            keep &= vals >= 0.0

        else:
            raise ValueError(f"[filters][goal] unknown goal rule: {m.get('rule')}")

    return keep


def _iqr_start_row(layout: dict[str, Any], cfg: dict[str, Any], start_row: Any) -> int:
    """
    Return 0-based starting row for IQR filtering.

    MATLAB default:
      "after_geom_vars" -> first row after D and Ubase in the ORIGINAL DB layout
    """
    if isinstance(start_row, str) and start_row.lower() == "after_geom_vars":
        d_rows = int(layout["D"]["nRows"])
        mbase = int(cfg["vars"]["Mbase"])
        return d_rows + mbase

    r = int(start_row)
    if r >= 1:
        return r - 1
    return r


def _iqr_keep_mask(
    db_current: Array,
    cfg: dict[str, Any],
    layout: dict[str, Any],
) -> BoolArray:
    """
    Legacy row-by-row IQR outlier filter on currently kept columns.

    A column is dropped if, for ANY selected row, its value lies outside:
      [Q1 - k*IQR, Q3 + k*IQR]

    This mirrors the MATLAB intent of a cumulative row-wise physical/outlier filter.
    """
    filters_cfg = cfg["filters"]
    iqr_cfg = dict(filters_cfg.get("iqr", {}) or {})
    debug_cfg = dict(filters_cfg.get("debug", {}) or {})

    k_iqr = float(iqr_cfg.get("k", 3.0))
    start_row = _iqr_start_row(layout, cfg, iqr_cfg.get("start_row", "after_geom_vars"))

    nr = db_current.shape[0]
    if start_row < 0:
        start_row = 0
    if start_row >= nr:
        return np.ones(db_current.shape[1], dtype=bool)

    keep = np.ones(db_current.shape[1], dtype=bool)

    for r in range(start_row, nr):
        vals = db_current[r, :]
        #q1 = float(np.quantile(vals, 0.25))
        #q3 = float(np.quantile(vals, 0.75))
        q1 = float(np.percentile(vals, 25, method="midpoint"))
        q3 = float(np.percentile(vals, 75, method="midpoint"))
        iqr = q3 - q1

        lo = q1 - k_iqr * iqr
        hi = q3 + k_iqr * iqr

        row_keep = (vals >= lo) & (vals <= hi)
        keep &= row_keep

        if bool(debug_cfg.get("iqr_rows", False)):
            dropped = int(np.sum(~row_keep))
            print(
                f"[filters][iqr][debug] row={r+1}, q1={q1:.6g}, q3={q3:.6g}, "
                f"iqr={iqr:.6g}, lo={lo:.6g}, hi={hi:.6g}, dropped={dropped}"
            )

    return keep


def apply_filters(
    db: Array,
    cfg: dict[str, Any],
    layout: dict[str, Any],
) -> FilterResult:
    """
    Apply column-wise filters in the same order as MATLAB:

      1) remove_nan
      2) goal
      3) iqr

    Returns:
      db_used = db[:, mask]
      info    = step-by-step counts
      mask    = boolean mask on ORIGINAL DB columns

    Logging matches MATLAB style.
    """
    db = np.asarray(db, dtype=float)
    ns = db.shape[1]

    cfg = _cfg_filters_defaults(cfg)

    mask = np.ones(ns, dtype=bool)
    info: dict[str, Any] = {"Ns_total": int(ns)}

    # =====================================================
    # (1) NaN filter
    # =====================================================
    if bool(cfg["filters"].get("remove_nan", True)):
        in_step = int(np.sum(mask))
        ok = np.all(np.isfinite(db[:, mask]), axis=0)
        kept_step = int(np.sum(ok))

        idx = np.flatnonzero(mask)
        mask[idx[~ok]] = False

        _force_keep_baseline(mask, cfg)

        info["nan_enable"] = True
        info["nan_in"] = in_step
        info["nan_kept"] = kept_step
        info["nan_dropped"] = in_step - kept_step

        _print_step("nan", kept_step, in_step, int(np.sum(mask)), ns)
    else:
        info["nan_enable"] = False
        info["nan_in"] = int(np.sum(mask))
        info["nan_kept"] = int(np.sum(mask))
        info["nan_dropped"] = 0

        _force_keep_baseline(mask, cfg)
        print(f"[filters][ nan] disabled (cumulative {int(np.sum(mask))} / {ns})")

    # =====================================================
    # (2) Goal filter
    # =====================================================
    goal_cfg = dict(cfg["filters"].get("goal", {}) or {})
    if bool(goal_cfg.get("enable", False)):
        in_step = int(np.sum(mask))

        c_rows = int(layout.get("C", {}).get("nRows", 0))
        if c_rows == 0:
            info["goal_enable"] = True
            info["goal_skipped"] = True
            info["goal_in"] = in_step
            info["goal_kept"] = in_step
            info["goal_dropped"] = 0

            _force_keep_baseline(mask, cfg)
            print(f"[filters][goal] skipped (C empty), cumulative {int(np.sum(mask))} / {ns}")
        else:
            metrics = goal_cfg.get("metrics", [])
            if not metrics:
                info["goal_enable"] = True
                info["goal_skipped"] = True
                info["goal_in"] = in_step
                info["goal_kept"] = in_step
                info["goal_dropped"] = 0

                _force_keep_baseline(mask, cfg)
                print(f"[filters][goal] skipped (metrics empty), cumulative {int(np.sum(mask))} / {ns}")
            else:
                idx_kept = np.flatnonzero(mask)
                db_cur = db[:, idx_kept]

                keep_local = _goal_keep_mask(db_cur, cfg, layout)
                kept_step = int(np.sum(keep_local))

                mask[idx_kept[~keep_local]] = False
                _force_keep_baseline(mask, cfg)

                info["goal_enable"] = True
                info["goal_skipped"] = False
                info["goal_in"] = in_step
                info["goal_kept"] = kept_step
                info["goal_dropped"] = in_step - kept_step

                _print_step("goal", kept_step, in_step, int(np.sum(mask)), ns)
    else:
        info["goal_enable"] = False
        info["goal_skipped"] = False
        info["goal_in"] = int(np.sum(mask))
        info["goal_kept"] = int(np.sum(mask))
        info["goal_dropped"] = 0

        _force_keep_baseline(mask, cfg)
        print(f"[filters][goal] disabled (cumulative {int(np.sum(mask))} / {ns})")

    # =====================================================
    # (3) IQR filter
    # =====================================================
    iqr_cfg = dict(cfg["filters"].get("iqr", {}) or {})
    if bool(iqr_cfg.get("enable", False)):
        in_step = int(np.sum(mask))

        idx_kept = np.flatnonzero(mask)
        db_cur = db[:, idx_kept]

        keep_local = _iqr_keep_mask(db_cur, cfg, layout)
        kept_step = int(np.sum(keep_local))

        mask[idx_kept[~keep_local]] = False
        _force_keep_baseline(mask, cfg)

        info["iqr_enable"] = True
        info["iqr_in"] = in_step
        info["iqr_kept"] = kept_step
        info["iqr_dropped"] = in_step - kept_step

        _print_step(" iqr", kept_step, in_step, int(np.sum(mask)), ns)
    else:
        info["iqr_enable"] = False
        info["iqr_in"] = int(np.sum(mask))
        info["iqr_kept"] = int(np.sum(mask))
        info["iqr_dropped"] = 0

        _force_keep_baseline(mask, cfg)
        print(f"[filters][ iqr] disabled (cumulative {int(np.sum(mask))} / {ns})")

    # =====================================================
    # final
    # =====================================================
    info["total_kept"] = int(np.sum(mask))
    db_used = db[:, mask]

    print(f"[filters] total kept {int(np.sum(mask))} / {ns} samples")

    return FilterResult(
        db_used=np.asarray(db_used, dtype=float),
        info=info,
        mask=np.asarray(mask, dtype=bool),
    )
